### PR TITLE
Update @yaireof/tagify for newly added exports field in package.json

### PR DESCRIPTION
**Pull request by blutorange**
Related

* https://github.com/yairEO/tagify/issues/1364
* https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/70285

[As I mentioned](https://github.com/yairEO/tagify/issues/1364#issuecomment-2286130016), tagify recently added an `exports` field to their package.json. This PR adds the exports fields to the types package as well, and adds a few tests for each possibly import path.

One note: I noticed some other packages use a separate `index.d.cts` for the common JS exports and a `index.d.ts` for the ESM exports (which mostly just re-exports from `index.d.cts`). I tried this as well, but I don't think we can get that to work without breaking existing code or duplicating code. We can investigate this for the next major version (unless somebody has an idea).

Also makes the following adjustments for version 4.27:

* Add `this.settings.classNames.dropdownItemSelected`, see https://github.com/yairEO/tagify/releases/tag/v4.27.0
* Add `tagify.setRangeAtStartEnd` and `tagify.setPlaceholder`, see https://github.com/yairEO/tagify/releases/tag/v4.26.6

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes:  https://github.com/yairEO/tagify/commit/d12daa63875af9df1f315ba6dc748c75ffc9c3cf
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/yaireo__tagify/package.json

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e72ddeecad6013bfc8f384db21bc513426a66d7c
</p>

File: types/yaireo__tagify/dist/tagify.esm.d.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// The package JSON "exports" field just redirects "./dist/tagify.esm.js" to "./index.d.ts"
+// This file exists solely for supporting Node 10, which does not support "exports" in package.json
+
+// Type tests are never run against a Node 10 environment.
+// Also, there's a lint rule that disallows relative imports in tests.
+// As a result, tests are never run against this file
+
+import Tagify = require("../index.js");
```
Comment (Brakebein) [inline types/yaireo__tagify/dist/tagify.esm.d.ts:8]: I don't get it completely. So, this file not referenced in the package.json exports. Tests are not run against file. But it is still needed somehow?

File: types/yaireo__tagify/test/dist/react.tagify.jsx.tsx
Comment (Brakebein) [inline types/yaireo__tagify/test/dist/react.tagify.jsx.tsx:1]: And why is this file needed with redundant tests?

File: types/yaireo__tagify/test/dist/tagify.min.js.ts
Comment (Brakebein) [inline types/yaireo__tagify/test/dist/tagify.min.js.ts:1]: Same question here?

**Review by Brakebein — COMMENTED** [review]
Thanks a lot! I was aware of that there were changes with the package.json exports, but didn't find any time to update this package.

I'm still struggling to understand why we need these extra test files. The package.json exports only reference `index.d.ts` and `dist/react.tagify.d.ts`. Why does it need the other test files?

File: types/yaireo__tagify/dist/tagify.esm.d.ts (hunk 1 above)
Comment (blutorange) [reply types/yaireo__tagify/dist/tagify.esm.d.ts:8]: Try removing this file, then run `pnpm test yaireo__tagify`. You'll get the following:

image

As far as I understand it, this [repo uses attw](https://github.com/DefinitelyTyped/DefinitelyTyped?tab=readme-ov-file#arethetypeswrongcli-attw-checks) to check that types exist for all possible entry points. It also contains checks for `node10` resolution mode. This mode does not support the `exports` field, so it falls back to raw paths as they are in the package.

File: types/yaireo__tagify/test/dist/react.tagify.jsx.tsx
Comment (blutorange) [reply types/yaireo__tagify/test/dist/react.tagify.jsx.tsx:1]: Perhaps we can combine some tests in a single file, but I would like to have at least a basic test for each possible export path declared in tagify's `package.json`. I.e. importing once from every possible path. To keep the redundancy low, I kept those tests very short, the full tests for all properties etc. are only in the main tests file.

Right now, most of these paths are mapped to the same type declarations file, but that might (and perhaps should) change in the future.

But I can reconsider, if you have another opinion on this?

```json
  "exports": {
    ".": {
      "import": "./dist/tagify.esm.js",
      "require": "./dist/tagify.js"
    },
    "./react": "./src/react.tagify",
    "./react.tagify": "./src/react.tagify",
    "./dist/react.tagify": "./dist/react.tagify.jsx",
    "./dist/react.tagify.jsx": "./dist/react.tagify.jsx",
    "./dist/tagify.min.js": "./dist/tagify.js",
    "./dist/tagify.min": "./dist/tagify.js",
    "./dist/tagify.esm.js": "./dist/tagify.esm.js",
    "./dist/tagify.polyfills.min.js": "./dist/tagify.polyfills.min.js",
    "./dist/tagify.vue": "./dist/tagify.vue",
    "./dist/tagify.css": "./dist/tagify.css"
  },
```

File: types/yaireo__tagify/dist/tagify.esm.d.ts (hunk 1 above)
Comment (Brakebein) [reply types/yaireo__tagify/dist/tagify.esm.d.ts:8]: Ok, I actually can't remember that attw was already part of the tests last time I updated the typings (or it just went through without any complains). So now, it's more than the basic type checking tests.

Would be nice, if attw would read the engine property of the original package (`"node": ">=16.15.0"`), than it could skip the node10 resolution mode. It's an already requested feature, but not yet capable of.

File: types/yaireo__tagify/test/dist/react.tagify.jsx.tsx
Comment (Brakebein) [reply types/yaireo__tagify/test/dist/react.tagify.jsx.tsx:1]: Ah now, I understood it: You are basically testing if each declared export resolves correctly. I'm fine with running only one basic test in these files. I would try to keep it as less redundant as possilbe. So currently, it's possible the closest it can get?

**Review by Brakebein — APPROVED** [review]
Now, as I understood the discussed points above, I'm fine with it.

File: types/yaireo__tagify/test/dist/react.tagify.jsx.tsx
Comment (blutorange) [reply types/yaireo__tagify/test/dist/react.tagify.jsx.tsx:1]: > You are basically testing if each declared export resolves correctly.

Exactly. And the basic test is to ensure it provides the correct exports.

> So currently, it's possible the closest it can get?

I mean, that's a bit subjective, but the basic test is basically just an import + one type import with one usage:

```ts
import Tagify = require("@yaireo/tagify/dist/tagify.esm.js");
import type { TagifySettings } from "@yaireo/tagify/dist/tagify.esm.js";
const settings: TagifySettings = {
    maxTags: 5,
};
const inputElement = document.createElement("input");
export const tagify = new Tagify(inputElement, settings);
```

As for the react tests, I shortened it a bit more so it's now only a single test for both components

```tsx
import * as React from "react";
import Tagify = require("@yaireo/tagify");
import Tags = require("@yaireo/tagify/dist/react.tagify");
import { MixedTags } from "@yaireo/tagify/dist/react.tagify";

export function TestMixedTagsMinimal(): React.ReactElement {
    return (
        <div>
            <MixedTags />
        </div>
    );
}

export function TestTagsTagifyRef(): React.ReactElement {
    const ref = React.useRef<Tagify>();
    return (
        <div>
            <Tags tagifyRef={ref} />
        </div>
    );
}
```

File: types/yaireo__tagify/dist/tagify.esm.d.ts (hunk 1 above)
Comment (blutorange) [reply types/yaireo__tagify/dist/tagify.esm.d.ts:8]: > Would be nice, if attw would read the engine property of the original package (`"node": ">=16.15.0"`), than it could skip the node10 resolution mode. It's an already requested feature, but not yet capable of.

It was also new for me, I had to check their readme. But I think I remember reading somewere that just declaring `"node": ">=16.15.0"` isn't enough. Apparently, when you set `moduleResolution: node` in a `tsconfig.json`, that's interpreted as `node10`, and there's still quite a few packages that do that.

But either way it feels like the build setup is not perfect yet -- skipping node 10 would be great. For now, that seems what we need to do to pass to not fail the build.

**Review by Brakebein — APPROVED** [review]

**Comment by blutorange** [comment]
Ready to merge
